[Unified Text Replacement] [macOS] Add SPI to WKWebViewConfiguration to configure unified text replacement behavior

rr-codes opened the pull request:
#### df94cb9a762a346ea7ef1cb57abd9f04774c2824
<pre>
[Unified Text Replacement] [macOS] Add SPI to WKWebViewConfiguration to configure unified text replacement behavior
<a href="https://bugs.webkit.org/show_bug.cgi?id=273886">https://bugs.webkit.org/show_bug.cgi?id=273886</a>
<a href="https://rdar.apple.com/127749267">rdar://127749267</a>

Reviewed by Wenson Hsieh.

* Source/WebKit/Shared/WebUnifiedTextReplacementContextData.h:
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::unifiedTextReplacementBehavior const):
(API::PageConfiguration::setUnifiedTextReplacementBehavior):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(convertToPlatformBehavior):
(convertToWebBehavior):
(-[WKWebViewConfiguration _setUnifiedTextReplacementBehavior:]):
(-[WKWebViewConfiguration _unifiedTextReplacementBehavior]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::selectionDidChange):
(WebKit::WebViewImpl::wantsCompleteUnifiedTextReplacementBehavior const):
(WebKit::WebViewImpl::unifiedTextReplacementBehavior const):
(WebKit::WebViewImpl::canHandleSwapCharacters const):
(WebKit::WebViewImpl::willBeginTextReplacementSession): Deleted.
(WebKit::WebViewImpl::didBeginTextReplacementSession): Deleted.
(WebKit::WebViewImpl::textReplacementSessionDidReceiveReplacements): Deleted.
(WebKit::WebViewImpl::textReplacementSessionDidUpdateStateForReplacement): Deleted.
(WebKit::WebViewImpl::didEndTextReplacementSession): Deleted.
(WebKit::WebViewImpl::textReplacementSessionDidReceiveTextWithReplacementRange): Deleted.
(WebKit::WebViewImpl::textReplacementSessionDidReceiveEditAction): Deleted.

Canonical link: <a href="https://commits.webkit.org/278617@main">https://commits.webkit.org/278617@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de623f27383b26811cfe6e52a0bb56e26c71f3ab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51142 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30444 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54399 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/1832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1506 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41649 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53241 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/28075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/44079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22765 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/1407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55995 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/26252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49048 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/27500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/44146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/28381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7425 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/27231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.h:
```diff
@@ -699,6 +703,10 @@ ALLOW_DEPRECATED_DECLARATIONS_END
 #endif
 
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+    WebKit::WebUnifiedTextReplacementBehavior unifiedTextReplacementBehavior() const;
```
```suggestion
    WebUnifiedTextReplacementBehavior unifiedTextReplacementBehavior() const;
```

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4567,7 +4566,12 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+bool WebViewImpl::wantsCompleteUnifiedTextReplacementBehavior() const
+{
+    return isEditable() || unifiedTextReplacementBehavior() == WebKit::WebUnifiedTextReplacementBehavior::Complete;
```
```suggestion
    return isEditable() || unifiedTextReplacementBehavior() == WebUnifiedTextReplacementBehavior::Complete;
```

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6381,11 +6385,18 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
 #endif // HAVE(TRANSLATION_UI_SERVICES) && ENABLE(CONTEXT_MENUS)
 
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+WebKit::WebUnifiedTextReplacementBehavior WebViewImpl::unifiedTextReplacementBehavior() const
```
```suggestion
WebUnifiedTextReplacementBehavior WebViewImpl::unifiedTextReplacementBehavior() const
```

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6381,11 +6385,18 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
 bool WebViewImpl::canHandleSwapCharacters() const
 {
-    return webViewCanHandleSwapCharacters();
+    return webViewCanHandleSwapCharacters() && unifiedTextReplacementBehavior() != WebKit::WebUnifiedTextReplacementBehavior::None;
```
```suggestion
    return webViewCanHandleSwapCharacters() && unifiedTextReplacementBehavior() != WebUnifiedTextReplacementBehavior::None;
```

whsieh reviewed (approved):
r=me, assuming the merge conflict is trivial to resolve.

rr-codes committed:
[Unified Text Replacement] [macOS] Add SPI to WKWebViewConfiguration to configure unified text replacement behavior
https://bugs.webkit.org/show_bug.cgi?id=273886
rdar://127749267

Reviewed by Wenson Hsieh.

* Source/WebKit/Shared/WebUnifiedTextReplacementContextData.h:
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::unifiedTextReplacementBehavior const):
(API::PageConfiguration::setUnifiedTextReplacementBehavior):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(convertToPlatformBehavior):
(convertToWebBehavior):
(-[WKWebViewConfiguration _setUnifiedTextReplacementBehavior:]):
(-[WKWebViewConfiguration _unifiedTextReplacementBehavior]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::selectionDidChange):
(WebKit::WebViewImpl::wantsCompleteUnifiedTextReplacementBehavior const):
(WebKit::WebViewImpl::unifiedTextReplacementBehavior const):
(WebKit::WebViewImpl::canHandleSwapCharacters const):
(WebKit::WebViewImpl::willBeginTextReplacementSession): Deleted.
(WebKit::WebViewImpl::didBeginTextReplacementSession): Deleted.
(WebKit::WebViewImpl::textReplacementSessionDidReceiveReplacements): Deleted.
(WebKit::WebViewImpl::textReplacementSessionDidUpdateStateForReplacement): Deleted.
(WebKit::WebViewImpl::didEndTextReplacementSession): Deleted.
(WebKit::WebViewImpl::textReplacementSessionDidReceiveTextWithReplacementRange): Deleted.
(WebKit::WebViewImpl::textReplacementSessionDidReceiveEditAction): Deleted.

Canonical link: https://commits.webkit.org/278617@main
